sound/soc/sof/ipc.c

```diff
@@ -736,6 +736,44 @@ int snd_sof_dsp_mailbox_init(struct snd_sof_dev *sdev, u32 dspbox,
 }
 EXPORT_SYMBOL(snd_sof_dsp_mailbox_init);
 
+int snd_sof_ipc_valid(struct snd_sof_dev *sdev)
+{
+	struct sof_ipc_fw_ready *ready = &sdev->fw_ready;
+	struct sof_ipc_fw_version *v = &ready->version;
+
+	dev_info(sdev->dev,
+		 " Firmware info: version %d:%d:%d-%s\n",  v->major, v->minor,
+		 v->micro, v->tag);
+	dev_info(sdev->dev,
+		 " Firmware: ABI %d:%d:%d Kernel ABI %d:%d:%d\n",
+		 SOF_ABI_VERSION_MAJOR(v->abi_version),
+		 SOF_ABI_VERSION_MINOR(v->abi_version),
+		 SOF_ABI_VERSION_PATCH(v->abi_version),
+		 SOF_ABI_MAJOR, SOF_ABI_MINOR, SOF_ABI_PATCH);
+
+	if (SOF_ABI_VERSION_INCOMPATIBLE(SOF_ABI_VERSION, v->abi_version)) {
+		dev_err(sdev->dev, "error: incompatible FW ABI version\n");
+		return -EINVAL;
+	}
+
+	if (ready->debug.build) {
+		dev_info(sdev->dev,
+			 " Firmware debug build %d on %s-%s - options:\n"
+			 "  lock debug: %s\n"
+			 "  lock vdebug: %s\n",
+			 v->build, v->date, v->time,
+			 ready->debug.locks ? "on" : "off",
+			 ready->debug.locks_verbose ? "on" : "off");
+	}
+
+	/* only copy the fw_version into debugfs at first boot */
+	if (sdev->first_boot)
+		memcpy(&sdev->fw_version, v, sizeof(*v));
+
```

Comment:
if we only copy the fw version at first boot, then we should only check and log messages for the first boot as well?

Reply:
We actually don't need to take this copy since we already have a copy of the container in sdev. I will be starting incremental patches to take care of the other ABI parts next week so there will be more coming.

Reply:
I am thinking if we can handle fw_ready() ops at first_boot only, I believe that works for our current implementation as we don't have any difference in fw_ready messages between first_boot and those after resumes.